feat(Badge): pt.2 - Badge component implementation / styling

jmuzina opened the pull request:
Requires merging #304 first - this PR is built on top of it.

## Done

Migrates the [Badge](https://canonical.github.io/react-components/?path=/docs/components-badge--docs) component to Pragma.

Fixes https://warthogs.atlassian.net/browse/WD-26107

## QA

- Review [badge docs](https://badge--67be30a0b85c4e44b6744158.chromatic.com/?path=/docs/badge--docs)
- Compare badge features against [React Components badge](https://canonical.github.io/react-components/?path=/docs/components-badge--docs)

### PR readiness check

- [x] PR should have one of the following labels:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [x] PR title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) format. 
- [x] All packages define the required scripts in `package.json`:
  - [x] All packages: `check`, `check:fix`, and `test`.
  - [x] Packages with build steps: `build` to build the package for development or distribution, `build:all` to build **all** artifacts. See [CONTRIBUTING.md](../old/CONTRIBUTING.md#24-full-artifact-builds-buildall) for details. 


https://github.com/user-attachments/assets/38d2bd73-8043-40e5-9475-678fb4b4029a

dgtlntv commented:
> @dgtlntv Can you have a look at the badge semantic colors in the [visual test](https://www.chromatic.com/review?appId=67be30a0b85c4e44b6744158&number=302&type=linked&view=activity) and let me know if they look as you would expect?

Hey sure! :)
Some things that stood out to me:
- I am not sure if this is because for the visual test a larger viewport was simulated and there was some rem scaling but badges, at least on the default view port, are supposed to have 14px font size and 16px line height. (If this is because of the visual test being generated on a larger viewport where font scaling is applied then I think it would be better if we moved to a viewport that is smaller or disable the font scaling because design primarily designs for the non font scaled version)
- the font weight is supposed to be a bit higher (so text remains readable at 12px font size). in the old (Vanilla) design it is 550. In the new typographic scale we would likely be moving to 600.
- And while the black text on the yellow background is correct I think it shows the short comings of the wcag contrast algorithm because according to apca white text, same as the other badges, white text has a higher contrast.

Other than that looks great 👍

jmuzina commented:
> the font weight is supposed to be a bit higher (so text remains readable at 12px font size). in the old (Vanilla) design it is 550. In the new typographic scale we would likely be moving to 600.

I have added a `font-weight-bold` token, set it to 600, and mapped the badge's font weight to this bold font weight token. The result:

<img width="151" height="84" alt="Screenshot 2025-08-20 at 17 01 14" src="https://github.com/user-attachments/assets/fb3741bf-2d05-48a5-8b00-f56ea193f7cc" />

jmuzina commented:
@dgtlntv Font size has been set to 0.75rem (12px) and line height to 1rem (16px), updated badge can be seen [here](https://67be30a0b85c4e44b6744158-ekfwrwltqk.chromatic.com/?path=/story/badge--large-number-rounded)

jmuzina commented on packages/react/ds-core/src/ui/Badge/styles.css:
```diff
@@ -0,0 +1,48 @@
+  padding: var(--badge-padding-vertical) var(--badge-padding-horizontal);
+  text-align: center;
+
+  &.positive {
+    background-color: var(--badge-color-background-positive);
+    color: var(--badge-color-text-positive);
+  }
+
+  &.negative {
+    background-color: var(--badge-color-background-negative);
+    color: var(--badge-color-text-negative);
+  }
+
+  &.caution {
+    background-color: var(--badge-color-background-caution);
+    color: var(--badge-color-text-caution);
+  }
+
+  &.information {
+    background-color: var(--badge-color-background-information);
+    color: var(--badge-color-text-information);
+  }
+}
```
@advl  Can you provide an example?

jmuzina replied:
Done with global intents

jmuzina commented on packages/react/ds-core/src/ui/Badge/Badge.stories.tsx:
```diff
@@ -0,0 +1,178 @@
+  component: Component,
+  parameters: {
+    docs: {
+      // TODO should we enable this for the entire package by default?
```
It seems it's not doable at the SB config level - it has to be enabled in `preview.ts` for each project

https://storybook.js.org/docs/writing-docs/code-panel

Here's a separate PR that performs that update
https://github.com/canonical/pragma/pull/312

jmuzina commented:
Don't merge added as the target branch is currently set to #304 to reduce the code diff surface for the ambitious early reviewer

Please review and merge #304 first

jmuzina commented:
Tomorrow I need to update this to add the types/utils package that @advl and I discussed today, so that the badge appearances can be mapped instead of hardcoded.

jmuzina commented on packages/react/ds-core/src/ui/Badge/Badge.Appearance.stories.tsx:
```diff
@@ -0,0 +1,82 @@
+
+type Story = StoryObj<typeof meta>;
+
+export const Positive: Story = {
```
I explored this:

```tsx
// Badge.Appearance.stories.tsx
import { SEVERITY } from "@canonical/ds-types";
  
SEVERITY.forEach(severityLevel => {
  const storyName = severityLevel.charAt(0).toUpperCase() + severityLevel.slice(1);
  // cannot use `export const` from within nested scope, or with dynamic names, so we can try with `module.exports`
  module.exports[storyName] = {
    args: {
      value: 10,
      appearance: severityLevel,
    },
    parameters: {
      docs: {
        description: {
          story: `The ${severityLevel} appearance indicates a ${severityLevel} status.`,
        },
      },
    },
  } as Story;
})
```

Howver, [storybook relies on stories being statically defined](https://github.com/storybookjs/storybook/issues/21687#issuecomment-1476253813) - this does not work. We would need to use Webpack, the [storybook's custom indexer feature](https://storybook.js.org/docs/api/main-config/main-config-indexers), or some other transpiler to accomplish this

The closest you could get is generating the stories with a mapping and then statically exporting them, which requires that you write some information from the stories in the export statement, which might be overly complex and too rigid. 

```tsx
// Assume this generates a mapping of severityName to Story using the Severity const array
const SeverityStories = generateSeverityStories();

export { Positive: SeverityStories.Positive, Negative: SeverityStories.Negative, Information: SeveirtyStories.Information } // and do on
```

An alternative might be statically creating the severity stories themselves, and taking advantage of CSF3 to spread most attributes of the default stories, changing only the appearance and anything else you need to on a per-story basis.

```tsx

export const Positive: Story = {
  args: {
    value: 42,
    appearance: "positive",
  },
  parameters: {
    docs: {
      description: {
        story:
          "The positive appearance uses green styling to indicate success, completion, or positive status.",
      },
    },
  },
};

export const Negative: Story = {
  args: {
    ...Positive.args,
    appearance: "negative",
  },
  parameters: {
    docs: {
      description: {
        story:
          "The negative appearance uses red styling to indicate errors, failures, or critical status.",
      },
    },
  },
};

export const Caution: Story = {
  args: {
    ...Positive.args,
   /// and so on...
```

jmuzina commented:
@advl I've updated this and #304 to reflect your suggestion to remove explicit clamping behavior, and I've adjusted the approach to global modifiers per our discussion today. Please have another look and let me know what you think.
